src/main/java/org/mongojack/internal/stream/JacksonCodec.java

```diff
@@ -30,4 +49,197 @@ public void encode(BsonWriter writer, T value, EncoderContext encoderContext) {
     public T decode(BsonReader reader, DecoderContext decoderContext) {
         return decoder.decode(reader, decoderContext);
     }
+
+    @Override
+    public T generateIdIfAbsentFromDocument(final T t) {
+        if (!documentHasId(t)) {
+            getWriter(t).accept(new BsonObjectId());
+        }
+        return t;
+    }
+
+    @Override
+    public boolean documentHasId(final T t) {
+        final BsonValue readValue = getDocumentId(t);
+        return readValue != null && !readValue.isNull();
+    }
+
+    @Override
+    public BsonValue getDocumentId(final T t) {
+        return getReader(t).get();
+    }
+
+    protected Supplier<BsonValue> getReader(final T t) {

```

Comment:
can you make this method private and the others that are only used in this file as well